### PR TITLE
refactor: Change snapshot loading interface

### PR DESCRIPTION
**Pull request by prathapsridharan**
## Reason for Change

- #5165 
- The entry point function `load_snapshot()` accepts a snapshot schema version and an optional snapshot id to enable loading snapshots are a particular snapshot schema version.

## Changes

- refactor the interface to `load_snapshot()` and the functions it calls to accept a snapshot schema version to load and an optional snapshot id to force load in the case the latest snapshot at a particular schema version is not needed

- **NOTE**: This change is merely a change to the interface. There are no changes to the behavior (ie output and side effects) of the functions being refactored. That change will come when we address #5166 and #5167 

## Testing steps

- Run unit tests locally and ensure that they all pass
- The following `rdev` functionally shows that the snapshot can be uploaded when you go to the gene expression app: https://ps-wmg-issue-5165-rdev-frontend.rdev.single-cell.czi.technology/gene-expression. The response of `\v2\query` is:

```
{
  "expression_summary": {},
  "snapshot_id": "1689721444",
  "term_id_labels": {
    "cell_types": {
      "CL:0000082 (cell culture)": {
        "CL:0000010": {
          "aggregated": {
            "cell_type_ontology_term_id": "CL:0000010",
            "name": "cultured cell",
            "order": 0,
            "total_count": 130626
          }
        }
```

- [Logs](https://us-west-2.console.aws.amazon.com/cloudwatch/home?region=us-west-2#logsV2:log-groups/log-group/$252Fdp$252Frdev$252Fps-wmg-issue-5165-rdev$252Fbackend/log-events/fargate$252Fweb$252F39b8a7c5901549fda8617e4a168b5142$3FfilterPattern$3D$2522Loading+WMG+snapshot$2522) from `cloudwatch` for the rdev prints the snapshot that was loaded:

```
2023-07-31T12:55:20.722-07:00
{
    "levelname": "INFO",
    "asctime": "2023-07-31T19:55:20.031Z",
    "name": "wmg",
    "message": "Loading WMG snapshot at s3://env-rdev-wmg/ps-wmg-issue-5165-rdev/1689721444",
    "lineno": 189,
    "pathname": "/single-cell-data-portal/backend/wmg/data/snapshot.py",
    "request_id": "ba1da4f0-82cd-4ff3-a801-08c992f77050"
}
```

### CODE REVIEW

File: backend/wmg/api/__init__.py
Comment (atolopko-czi) [inline backend/wmg/api/__init__.py:1]: nit: would put these constants in an explicit source file (`config.py`, `contants.py`, `globals.py`, etc), not `__init__.py`. I believe the convention is to only put package loading-related behavior and docs in this file.

**Review by atolopko-czi — COMMENTED** [review]
Gave it quick once over, and it all seems reasonable. A few minor comments. 

What do we make of all the code coverage warnings? Though valid, they seem to have popped up unnecessarily...

Will recuse myself as an approver, since Viz team members are better suited to approving! :)

**Review by atarashansky — APPROVED** [review]

File: backend/wmg/api/__init__.py
Comment (prathapsridharan) [reply backend/wmg/api/__init__.py:1]: @atolopko-czi - I will move to an explicit source file. I will do the same for the writer too.

File: backend/wmg/api/v2.py [hunk 1]
```diff
@@ -26,7 +27,11 @@
 
 
 def primary_filter_dimensions():
-    snapshot: WmgSnapshot = load_snapshot()
+    snapshot: WmgSnapshot = load_snapshot(
+        snapshot_schema_version=WMG_API_SNAPSHOT_SCHEMA_VERSION,
+        specific_snapshot_id_to_load=WMG_API_FORCE_LOAD_SNAPSHOT_ID,
```
[the remark this answers is not in the document]
Comment (prathapsridharan) [reply backend/wmg/api/v2.py:32]: @atolopko-czi - I will take this suggestion too

File: backend/wmg/data/snapshot.py [hunk 2]
```diff
@@ -98,33 +105,94 @@ def build_dataset_metadata_dict(self):
...
-    been updated.
-    @return: WmgSnapshot object
+    Loads and caches the snapshot identified by the snapshot schema version and a snapshot id.
+
+    By default, this functions loads the latest snapshot id for a given schema version.
+    If a specific snapshot id is given, it will load that snapshot id. Note, that if a specific
+    snapshot id is given, it should be associated with the given schema version.
+
+    The snapshot representation is cached in memory. Therefore, multiple calls to this function
+    will not simply return the cached snapshot if there isn't a newer snapshot id.
```
[the remark this answers is not in the document]
Comment (prathapsridharan) [reply backend/wmg/data/snapshot.py:117]: @atolopko-czi - will fix this

**Comment by prathapsridharan** [comment]
@atolopko-czi 

> What do we make of all the code coverage warnings? Though valid, they seem to have popped up unnecessarily...

I will look into this but my changes are unrelated to these warnings. Maybe they have been showing up for a while and we didn't notice them?

**Comment by atolopko-czi** [comment]
> @atolopko-czi
> 
> > What do we make of all the code coverage warnings? Though valid, they seem to have popped up unnecessarily...
> 
> I will look into this but my changes are unrelated to these warnings. Maybe they have been showing up for a while and we didn't notice them?

Agreed, they are unrelated to the PR, so no worries. Was only hoping you understand cause more than I did. 😄

**Pull request merged**